fix: Delete PVC preparation pods if a PVC failure event is detected

dkwon17 opened the pull request:
Signed-off-by: David Kwon <dakwon@redhat.com>

<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests
-->

### What does this PR do?
This PR fixes eclipse/che#20216 is an alternative PR to https://github.com/eclipse-che/che-server/pull/76.

In more detail, this PR watches for pod events for rm pods (i.e. PVC preparation pods). The following pod events are the events that happen when the common PVC is terminating: `Failed`, `FailedScheduling`, `FailedMount`. If these pod events are detected, the rm pod is deleted.

### Screenshot/screencast of this PR
<!-- Please include a screenshot or a screencast explaining what is doing this PR -->
When deleting 20 workspaces, we see the rm pods and PVC terminate without and rm pods stuck in the `Pending` state.
![che-20216](https://user-images.githubusercontent.com/83611742/128775163-182c51a3-7ab7-48b6-b787-dcba08038a03.gif)


### What issues does this PR fix or reference?
<!-- Please include any related issue from eclipse che repository (or from another issue tracker).
     Include link to other pull requests like documentation PR from [the docs repo](https://github.com/eclipse/che-docs)
-->
eclipse/che#20216 

### How to test this PR?
<!-- Please explain for example :
  - The test platform (openshift, kubernetes, minikube, CodeReady Container, docker-desktop, etc)
  - Installation method: chectl / che-operator
  - steps to reproduce
 -->
The image: `quay.io/dkwon17/che-server:che-20216-delete`

1. Deploy on minikube:
```
$ minikube start --addons=ingress --vm=true --memory=8192
$ chectl server:deploy --installer operator -p minikube --cheimage=quay.io/dkwon17/che-server:che-20216-delete
```

2. Login using `chectl`:
```
$ chectl auth:login
```

3. Create 20 workspaces:
```
$ for run in {1..20}; do chectl workspace:create -f https://raw.githubusercontent.com/openshiftio/saas-openshiftio/master/devfile.yaml; done
```
4. Start and stop a workspace so that a common PVC is created.

5. Delete all 20 workspaces:
![image](https://user-images.githubusercontent.com/83611742/128253201-ea2ac9aa-c245-4a0d-b121-987a07a34f8f.png)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

dkwon17 commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java:
```diff
@@ -218,17 +225,23 @@ private void execute(
-            Arrays.toString(command),
-            Strings.nullToEmpty(logs).replace("\n", " \\n")); // Force logs onto one line
+
+      if (pvcExistsAndBound(namespace, pvcName)) {
```
The PR also checks if PVC exists and is bound before creating the pod. This is because the PVC could have been in the processing of terminating before `execute()` is run.

Also it seems that because of this `if` condition, only 1/9 tests in PVCSubPathHelperTest.java succeed. I will update/add tests very soon.

sleshchenko replied:
it's going to fail in WaitForFirstConsumer storage type.
it's default for RHPDS cluster and pretty common I think.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java:
```diff
@@ -218,16 +227,21 @@ private void execute(
-            Arrays.toString(command),
-            Strings.nullToEmpty(logs).replace("\n", " \\n")); // Force logs onto one line
+
+      if (pvcExistsAndBound(namespace, pvcName)) {
```
anyway I would not increase code nesting and go in the following way:
```java
if (!doesPvcExist(pvc)) {
    // no PVC exist, so nothing to clean up
	return;
}

// do clean up
```
and yeah, this check could be done much earlier even before pod object creation, or even on other more high level class.

ibuziuk replied:
not only should we check that PVC does not exist, but also that it is not in a terminating state.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java:
```diff
@@ -247,6 +261,56 @@ private void execute(
+   *
+   * @param deployments the KubernetesDeployments object
+   * @param podName the name of the Pod to check events for
+   * @throws InfrastructureException
```
it may be better documented

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java:
```diff
@@ -247,6 +261,56 @@ private void execute(
+   */
+  private void watchFailureEventAndTerminatePod(KubernetesDeployments deployments, String podName)
+      throws InfrastructureException {
+    deployments.watchEvents(
```
I'm not sure who is the author of watchEvents, maybe even me... But AFAIU this event watcher instance is going to live forever, which would lead to more memory consumption and making server slower to process. Please consider contribution unsubscribe method to deployments where you pass watcher instance... Or make deployments return subscriptions where you can invoke unsubscribe in finally statement.

dkwon17 replied:
The watch is being cleaned up here in the finally block:
https://github.com/eclipse-che/che-server/blob/90968896473bf0172d9dc143906a91b9f89c96aa/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java#L253-L255

I think this is ok?

sleshchenko replied:
Huh, glad that it's in place!
I must to finish my cup of coffee before reviewing PR :-D

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java:
```diff
@@ -247,6 +261,56 @@ private void execute(
+            try {
+              LOG.debug(
+                  "Deleting pod: '{}' due to failure event: '{}'", podName, event.getMessage());
+              deployments.delete(event.getPodName());
```
Are you sure it does not conflict with
```java
        final Pod finished = deployments.wait(podName, WAIT_POD_TIMEOUT_MIN, POD_PREDICATE::apply);
```
?

I mean: please make sure wait is interrupted after you remove the pod.

dkwon17 replied:
Sounds good. To interrupt the wait, I thought maybe `deployments.wait()` could return a `CompletableFuture<Pod>`, and call `cancel()` to interrupt the wait if needed, but I'm having trouble figuring out how to call `watch.close()` after canceling the `CompletableFuture`. I suppose `deployments.wait()` could return a Java object containing both the `CompletableFuture` and the `Watch` object, but I'm not sure if that's the way to go.
https://github.com/eclipse-che/che-server/blob/90968896473bf0172d9dc143906a91b9f89c96aa/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java#L236

Maybe I should instead create a method for `KubernetesDeployments` called something like `watchEventsAndWaitPod`, that would both watch for failure events and return if any, and, wait/watch for pod completion?

sleshchenko reviewed (commented):
Some corner cases still need attention.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java:
```diff
@@ -247,6 +261,56 @@ private void execute(
+        });
+  }
+
+  private boolean isPodFailureEvent(PodEvent event) {
+    return event.getReason().equals(POD_EVENT_FAILED)
+        || event.getReason().equals(POD_EVENT_FAILED_SCHEDULING)
+        || event.getReason().equals(POD_EVENT_FAILED_MOUNT);
+  }
```
This has some overlap with [`UnrecoverablePodEventListenerFactory`](https://github.com/eclipse-che/che-server/blob/main/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/UnrecoverablePodEventListenerFactory.java) -- I wonder if there's some way to reuse the functionality (maybe not, since the unrecoverable events there are configured by property)

ibuziuk replied:
As far as I recall, unrecorerable events kick in only during the workspace startup phase, async removal of dirs in PVC is a separate routine

dkwon17 replied:
Also, one of my concerns about watching for these failure events is that: a terminating PVC would cause failure events for the rm-pods, but, observing failure events might not necessarily mean that its caused by a terminating PVC? Maybe there could be other reasons (that are not PVC-related) why these rm pods might have failure events?

dkwon17 commented:
Thank you everyone, I have made some new changes to address the current feedback.

ibuziuk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:
```diff
@@ -235,17 +235,17 @@ public void prepare(
-      return;
-    }
-
     AccountImpl account = ((WorkspaceImpl) workspace).getAccount();
     if (isPersonalAccount(account) && accountHasNoWorkspaces(account)) {
       log.debug("Deleting the common PVC: '{}',", configuredPVCName);
       deleteCommonPVC(workspace);
       return;
     }
 
+    if (EphemeralWorkspaceUtility.isEphemeral(workspace)) {
+      return;
+    }
+
```
@dkwon17 could we have a test case for this part e.g. if the workspace is ephemeral and account Has No Workspaces the deleteWorkspacePVC will be called?

dkwon17 replied:
Sounds good, my new commit adds this test: `shouldDeleteCommonPVCIfPersistAttributeIsSetToFalseInWorkspaceConfigAndWorkspacesEmptyWhenCleanupCalled` (maybe the name is too long?)

ibuziuk commented:
[crw-ci-test --rebuild]

ibuziuk commented:
[crw-ci-test]

dmytro-ndp commented:
[crw-ci-test --rebuild]

ibuziuk commented:
[crw-ci-test --rebuild]

ibuziuk commented:
[crw-ci-test --rebuild]

ibuziuk commented:
[crw-ci-test --rebuild]

ibuziuk commented:
[crw-ci-test --rebuild]

skabashnyuk commented:
please. remove all lenient() from the body of the test.
Usually, that means that you add a useless stub or removed a code that used that stub.

dkwon17 commented:
Sounds great, I have removed the `lenient()` calls in the test bodies, but kept the ones I've added in the `setup` method in `PVCSubPathHelperTest`:
https://github.com/eclipse-che/che-server/blob/ae3d35e4b9b763f6b40ecedf2c75c634dc14ca6f/infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelperTest.java#L136-L141

dkwon17 commented:
This PR also removes `mkdir` pods when failure events happen. For example, this is the debug log when an `mkdir` pod has a`MountVolume.SetUp failed for volume "default-token-plrpl"` event.
```
2021-08-19 14:51:17,271[//10.96.0.1/...]  [DEBUG] [e.c.w.i.k.n.p.PVCSubPathHelper 321]  - Deleting pod: 'mkdir-workspacecmadh8rlbaddjaji' due to failure event: 'MountVolume.SetUp failed for volume "default-token-plrpl" : failed to sync secret cache: timed out waiting for the condition'
2021-08-19 14:51:17,274[aceSharedPool-4]  [DEBUG] [f.k.c.d.i.AbstractWatchManager 195]  - Force closing the watch io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager@1bce17d9
2021-08-19 14:51:17,277[aceSharedPool-4]  [DEBUG] [i.f.kubernetes.client.Watcher 36]    - Watcher closed
2021-08-19 14:51:17,277[aceSharedPool-4]  [DEBUG] [k.c.d.i.WatchConnectionManager 45]   - Closing websocket okhttp3.internal.ws.RealWebSocket@14219a7a
2021-08-19 14:51:17,279[aceSharedPool-4]  [ERROR] [e.c.w.i.k.n.p.PVCSubPathHelper 250]  - Unable to perform '[mkdir, -m, 777, -p, /tmp/job_mount/workspacecmadh8rlbaddjaji/plugins/, /tmp/job_mount/workspacecmadh8rlbaddjaji/theia-local/, /tmp/job_mount/workspacecmadh8rlbaddjaji/che-logs-che-workspace-pod/, /tmp/job_mount/workspacecmadh8rlbaddjaji/projects/]' command for the workspace 'workspacecmadh8rlbaddjaji' cause: 'Cancelled waiting for pod: 'mkdir-workspacecmadh8rlbaddjaji''
``` 

However, fixing eclipse/che#20216 only needs `rm` pods to be deleted when failure events (caused by terminating PVC) occur.

Would it be a good idea to update this PR to watch for failure events for only `rm` pods?

ibuziuk commented:
> Would it be a good idea to update this PR to watch for failure events for only rm pods?

@dkwon17 yes, let's do it only for rm pods. In general I belive on the new versions of k8s mkdir routine is not going to be required - https://github.com/eclipse/che/issues/19027#issuecomment-787757279

dkwon17 commented:
@ibuziuk sounds great, I have added a commit where failure events are watched only for rm pods.
